File: apps/mobile/test/acp_host_ui_features_test.dart
import 'package:eco_mobile/core/models/acp_host_ui_features.dart';
import 'package:eco_mobile/core/models/thread_models.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  test('fromJson missing object is all show', () {
    expect(AcpHostUiFeatures.fromJson(null).showContextUsage, isTrue);
    expect(AcpHostUiFeatures.fromJson(null).showBilling, isTrue);
  });

  test('fromJson keeps valid hide and falls dirty columns back to show', () {
    final features = AcpHostUiFeatures.fromJson({
      'contextUsage': 'hide',
      'billing': 'nope',
    });
    expect(features.showContextUsage, isFalse);
    expect(features.showBilling, isTrue);
  });

  test('ThreadSummary.fromJson reads hostUiFeatures', () {
    final thread = ThreadSummary.fromJson({
      'id': 'thr_1',
      'title': 't',
      'prompt': 'p',
      'workspacePath': '/tmp',
      'status': 'idle',
      'createdAt': '',
      'updatedAt': '',
      'message': '',
      'coreKind': 'acp',
      'hostUiFeatures': {'contextUsage': 'hide', 'billing': 'hide'},
    });
    expect(thread.hostUiFeatures.showContextUsage, isFalse);
    expect(thread.hostUiFeatures.showBilling, isFalse);
  });
}

File: packages/runtime/test/acp-host-ui-features.test.ts
import { expect, test } from "bun:test";
import {
  DEFAULT_ACP_HOST_UI_FEATURES,
  isAcpHostUiFeatureVisible,
  normalizeAcpHostUiFeatures,
  resolveAcpHostUiFeatures,
} from "../src/acp-host-ui-features.js";

test("cursor hides context usage and billing", () => {
  expect(resolveAcpHostUiFeatures({ coreKind: "acp", acpAgentId: "cursor" })).toEqual({
    contextUsage: "hide",
    billing: "hide",
  });
});

test("non-acp cores show both columns", () => {
  for (const coreKind of ["claude", "codex", "pi"] as const) {
    expect(resolveAcpHostUiFeatures({ coreKind })).toEqual(DEFAULT_ACP_HOST_UI_FEATURES);
  }
});

test("unknown or missing acp agent defaults to show", () => {
  expect(resolveAcpHostUiFeatures({ coreKind: "acp" })).toEqual(DEFAULT_ACP_HOST_UI_FEATURES);
  expect(resolveAcpHostUiFeatures({ coreKind: "acp", acpAgentId: "other" })).toEqual(
    DEFAULT_ACP_HOST_UI_FEATURES,
  );
  expect(resolveAcpHostUiFeatures({ coreKind: "acp", acpAgentId: "  " })).toEqual(
    DEFAULT_ACP_HOST_UI_FEATURES,
  );
});

test("normalize missing object is all show", () => {
  expect(normalizeAcpHostUiFeatures(undefined)).toEqual(DEFAULT_ACP_HOST_UI_FEATURES);
  expect(normalizeAcpHostUiFeatures(null)).toEqual(DEFAULT_ACP_HOST_UI_FEATURES);
});

test("normalize dirty column falls back to show and keeps a valid hide", () => {
  expect(
    normalizeAcpHostUiFeatures({ contextUsage: "hide", billing: "nope" }),
  ).toEqual({ contextUsage: "hide", billing: "show" });
  expect(normalizeAcpHostUiFeatures({ contextUsage: "hide" })).toEqual({
    contextUsage: "hide",
    billing: "show",
  });
});

test("isAcpHostUiFeatureVisible treats missing features as show", () => {
  expect(isAcpHostUiFeatureVisible(undefined, "billing")).toBe(true);
  expect(
    isAcpHostUiFeatureVisible({ contextUsage: "hide", billing: "hide" }, "billing"),
  ).toBe(false);
});
